### PR TITLE
Add TLSEventSupport

### PR DESCRIPTION
**Pull request by maskit**
This is preparation to support [TLS User Agent Hooks](https://docs.trafficserver.apache.org/en/latest/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.html) on QUIC connections. The goal on this PR is, like other TLS*Support mix-ins, to remove dynamic_casts to SSLNetVC. Wiring for QUICNetVC will be made later on another PR.

### CODE REVIEW

File: include/iocore/net/TLSEventSupport.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+protected:
+  void clear();
+
+  SSLHandshakeHookState get_handshake_hook_state();
+  void                  set_handshake_hook_state(SSLHandshakeHookState state);
+  bool                  is_invoked_state() const;
+  int                   invoke_tls_event();
+  void                  resume_tls_event();
+
+  virtual bool _is_tunneling_requested()   = 0;
```
Comment (brbzull0) [inline include/iocore/net/TLSEventSupport.h:75]: I would make it `const`.

File: src/iocore/net/SSLNetVConnection.cc [hunk 2]
```diff
@@ -1581,28 +1465,18 @@ SSLNetVConnection::sslClientHandShakeEvent(int &err)
...
     }
 
-    sslHandshakeHookState = HANDSHAKE_HOOKS_OUTBOUND_PRE;
+    this->set_handshake_hook_state(TLSEventSupport::SSLHandshakeHookState::HANDSHAKE_HOOKS_OUTBOUND_PRE);
   }
 
   // Do outbound hook processing here
   // Continue on if we are in the invoked state.  The hook has not yet reenabled
-  if (sslHandshakeHookState == HANDSHAKE_HOOKS_OUTBOUND_PRE_INVOKE) {
+  if (this->is_invoked_state()) {
```
Comment (brbzull0) [inline src/iocore/net/SSLNetVConnection.cc:1473]: is it relevant in which particular `INVOKE` state it is?

**Review by brbzull0 — COMMENTED** [review]
looks good, just one question though.

File: src/iocore/net/SSLNetVConnection.cc (hunk 2 above)
Comment (maskit) [reply src/iocore/net/SSLNetVConnection.cc:1473]: No, not really in terms of the flow. Like the function name suggests, whether it's one of `INVOKE`s is the key.

**Comment by brbzull0** [comment]
[approve ci autest]

**Review by brbzull0 — APPROVED** [review]
Thanks.

**Pull request merged**